### PR TITLE
Configurable population resources

### LINKED ISSUE #2408: Declare resources to be used to populate QuestionnaireResponse via ActionParameter
**Describe the feature request.**
The `QuestionnaireResponse` requires previously extracted Resources to populate its items e.g. when editing a questionnaire. This enhancement will allow us to use the existing [ActionParameter](https://github.com/opensrp/fhircore/blob/c7758643ac515a49a1bcd59b304c1173475062f5/android/engine/src/main/java/org/smartregister/fhircore/engine/domain/model/ActionConfig.kt#LL74C12-L74C27) to declare the resources to be used during population. 

**Additional context**
The current implementation is only limited to populate resources using the resource the questionnaire is linked to e.g. if the Questionnaire is for completing a Task, only the task resource will be used during population of `QuestionnaireResponse` items. The enhancement should allow us to provide references for other necessary resources via configuration. 

**Acceptance criteria**
Be able to configure the resources to be used to populate `QuestionnaireResponse` via the `ActionParameter` config.

**Implementation plan (For Engineers)**
1. Update [ActionParameter](https://github.com/opensrp/fhircore/blob/main/android/engine/src/main/java/org/smartregister/fhircore/engine/domain/model/ActionConfig.kt#LL74C12-L74C27) by adding an optional (nullable) `resourceType: ResourceType` property
2. Add a new type named QUESTIONNAIRE_RESPONSE_POPULATION_RESOURCE in [ActionParameterType](https://github.com/opensrp/fhircore/blob/main/android/engine/src/main/java/org/smartregister/fhircore/engine/domain/model/ActionParameterType.kt)
3. Extract all `ActionParameter` of the new type  (QUESTIONNAIRE_RESPONSE_POPULATION_RESOURCE) to a variable to be used in the `QuestionnaireResponse` population logic. `ActionParameter.value` should contain the resourceID. This resource ID will be used to retrieve the resource. The type of the resource should be provided via `ActionParameter.resourceType`
4. The update configs for launching questionnaires where population with resource is needed. E.g:
```json
"params": [
            {
              "paramType": "QUESTIONNAIRE_RESPONSE_POPULATION_RESOURCE",
              "resourceType": "Task"
              "value": "@{theComputedTaskId}"
            },
           {
              "paramType": "QUESTIONNAIRE_RESPONSE_POPULATION_RESOURCE",
              "resourceType": "Patient"
              "value": "@{theComputedPatientId}"
            }
          ]
```
5. Update the logic for populating `QuestionnaireResponse` to use the resources retrieved from the configured `ActionParameter`s

### PR DESCRIPTION
**Pull request by aurangzaibumer**
**IMPORTANT: Where possible all PRs must be linked to a Github issue**

Fixes #2408

**Engineer Checklist**
- [x] I have written **Unit tests** for any new feature(s) and edge cases for bug fixes
- [x] I have added any strings visible on UI components to the `strings.xml` file
- [x] I have updated the  [CHANGELOG.md](./CHANGELOG.md) file for any notable changes to the codebase
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the project's style guide
- [x] I have built and run the FHIRCore app to verify my change fixes the issue and/or does not break the app 
- [x] I have checked that this PR does NOT introduce **breaking changes** that require an update to **_Content_** and/or **_Configs_**? _If it does add a sample here or a link to exactly what changes need to be made to the content._

**Code Reviewer Checklist**
- [ ] I have verified **Unit tests** have been written for any new feature(s) and edge cases
- [ ] I have verified any strings visible on UI components are in the `strings.xml` file
- [ ] I have verifed the [CHANGELOG.md](./CHANGELOG.md) file has any notable changes to the codebase
- [ ] I have verified the solution has been implemented in a configurable and generic way for reuseable components
- [ ] I have built and run the FHIRCore app to verify the change fixes the issue and/or does not break the app

### CODE REVIEW

**Comment by FikriMilano** [comment]
currently enhancing the test

File: android/quest/src/main/java/org/smartregister/fhircore/quest/ui/questionnaire/QuestionnaireActivity.kt [hunk 1]
```diff
@@ -178,6 +180,15 @@ open class QuestionnaireActivity : BaseMultiLanguageActivity(), View.OnClickList
...
   }
 
+  @VisibleForTesting
+  internal fun getResourcesFromParamsForQR(): Map<ResourceType?, String> {
+    val resourceMap =
+      actionParams
+        .filter { it.paramType == ActionParameterType.QUESTIONNAIRE_RESPONSE_POPULATION_RESOURCE }
+        .associate { it.resourceType to it.value }
+    return resourceMap
+  }
```
Comment (ellykits) [inline android/quest/src/main/java/org/smartregister/fhircore/quest/ui/questionnaire/QuestionnaireActivity.kt:190]: Can be converted to expression function to improve readability 
```suggestion
  internal fun getResourcesFromParamsForQR(): Map<ResourceType?, String>  =
      actionParams
        .filter { it.paramType == ActionParameterType.QUESTIONNAIRE_RESPONSE_POPULATION_RESOURCE }
        .associate { it.resourceType to it.value }
     
```

File: android/quest/src/test/java/org/smartregister/fhircore/quest/ui/questionnaire/QuestionnaireActivityTest.kt [hunk 2]
```diff
@@ -922,6 +923,58 @@ class QuestionnaireActivityTest : ActivityRobolectricTest() {
...
+
+    val questionnaireFragment = spyk<QuestionnaireFragment>()
+    every { questionnaireFragment.getQuestionnaireResponse() } returns QuestionnaireResponse()
+
+    val controller = Robolectric.buildActivity(QuestionnaireActivity::class.java, intent)
+    questionnaireActivity = controller.create().resume().get()
+
+    val resourceMap = questionnaireActivity.getResourcesFromParamsForQR()
+
+    Assert.assertEquals(1, resourceMap.size)
```
Comment (ellykits) [inline android/quest/src/test/java/org/smartregister/fhircore/quest/ui/questionnaire/QuestionnaireActivityTest.kt:975]: Can we add other verifications for this single item of the resourceMap?

Comment (ellykits) [reply android/quest/src/test/java/org/smartregister/fhircore/quest/ui/questionnaire/QuestionnaireActivityTest.kt:975]: Verify the paramType for instance

File: android/quest/src/main/assets/configs/app/profiles/household_profile_config.json [hunk 3]
```diff
@@ -565,6 +565,7 @@
             "setOrganizationDetails": true,
             "type": "EDIT",
             "resourceIdentifier": "@{familyLogicalId}",
+            "resourceType": "Group",
```
Comment (ellykits) [inline android/quest/src/main/assets/configs/app/profiles/household_profile_config.json:568]: Do we need to include the questionnaire response action params to this config?

File: android/quest/src/test/java/org/smartregister/fhircore/quest/ui/questionnaire/QuestionnaireViewModelTest.kt [hunk 4]
```diff
@@ -2026,48 +1945,41 @@ class QuestionnaireViewModelTest : RobolectricTest() {
...
-
+    assertTrue(result.hasItem())
     coVerify { fhirEngine.get(group.resourceType, group.id) }
-
     val slotPopulationResources = slot<ArrayList<Resource>>()
     coVerify {
       questionnaireViewModel.populateQuestionnaireResponse(any(), capture(slotPopulationResources))
     }
-    assertTrue(slotPopulationResources.captured.isNotEmpty())
+    assertTrue(slotPopulationResources.captured.isEmpty())
```
Comment (ellykits) [inline android/quest/src/test/java/org/smartregister/fhircore/quest/ui/questionnaire/QuestionnaireViewModelTest.kt:1983]: Have we written tests where the captured param is not Empty?

**Review by ellykits — CHANGES_REQUESTED** [review]

**Comment by ellykits** [comment]
@FikriMilano @aurangzaibumer Submit PRs on the client configs affected by this refactor on the fhir-resources repo.

**Comment by FikriMilano** [comment]
@ellykits 
I don't really know what client configs will be affected.
@dubdabasoduba can comment on this.

The implementation is backward compatible.

File: android/quest/src/main/java/org/smartregister/fhircore/quest/ui/questionnaire/QuestionnaireActivity.kt (hunk 1 above)
Comment (FikriMilano) [reply android/quest/src/main/java/org/smartregister/fhircore/quest/ui/questionnaire/QuestionnaireActivity.kt:190]: okay

File: android/quest/src/main/assets/configs/app/profiles/household_profile_config.json (hunk 3 above)
Comment (FikriMilano) [reply android/quest/src/main/assets/configs/app/profiles/household_profile_config.json:568]: 1. no, household_profile_config.json doesn't need that.
2. resourceType as Group, is to fix launchContext issue. Where if not declared as, by default it tries to load a Patient, and in this case with a Group ID

File: android/quest/src/test/java/org/smartregister/fhircore/quest/ui/questionnaire/QuestionnaireViewModelTest.kt (hunk 4 above)
Comment (FikriMilano) [reply android/quest/src/test/java/org/smartregister/fhircore/quest/ui/questionnaire/QuestionnaireViewModelTest.kt:1983]: yep, check these

- https://github.com/opensrp/fhircore/blob/7fbd1e41d2029edcb1cf5a4d3aada9a4bda968fb/android/quest/src/test/java/org/smartregister/fhircore/quest/ui/questionnaire/QuestionnaireViewModelTest.kt#L1738
- https://github.com/opensrp/fhircore/blob/7fbd1e41d2029edcb1cf5a4d3aada9a4bda968fb/android/quest/src/test/java/org/smartregister/fhircore/quest/ui/questionnaire/QuestionnaireViewModelTest.kt#L1820
- https://github.com/opensrp/fhircore/blob/7fbd1e41d2029edcb1cf5a4d3aada9a4bda968fb/android/quest/src/test/java/org/smartregister/fhircore/quest/ui/questionnaire/QuestionnaireViewModelTest.kt#L1901

**Review by ellykits — APPROVED** [review]

**Pull request merged**